source/common/router/router.cc

```diff
@@ -846,6 +857,30 @@ void Filter::sendNoHealthyUpstreamResponse() {
                              StreamInfo::ResponseCodeDetails::get().NoHealthyUpstream);
 }
 
+void Filter::giveUpRetryAndShadow() {
+  cluster_->trafficStats()->retry_or_shadow_abandoned_.inc();
+  retry_state_.reset();
+  active_shadow_policies_.clear();
+  request_buffer_overflowed_ = true;
+}
+
+void Filter::onUpstreamConnectionEstablished() {
+  ENVOY_LOG(trace, "Got connection with upstream.");
+  increased_retry_shadow_buffer_limit_until_got_connection_ = false;
+  if (!request_buffer_overflowed_ &&
+      getLength(callbacks_->decodingBuffer()) > retry_shadow_buffer_limit_) {
+    ENVOY_LOG(debug,
+              "The request payload has at least {} bytes data which exceeds buffer limit {}. Give "
+              "up on the retry/shadow.",
+              getLength(callbacks_->decodingBuffer()), retry_shadow_buffer_limit_);
+    // Disabling retry and shadow: the request is to big.
+    giveUpRetryAndShadow();
+    if (getLength(callbacks_->decodingBuffer()) > callbacks_->decoderBufferLimit()) {
+      callbacks_->setDecoderBufferLimit(getLength(callbacks_->decodingBuffer()));
```

Comment:
So the new bufferlimit right now is  < `retry_connection_failure_buffer_limit_` and >`retry_shadow_buffer_limit_`, other than that we will early local reply later, am I reading correctlly?

Reply:
that is correct